Freeze pb model will not remove is_training flag of bn attr and it is moreover still set as true

HwMohanLiu opened the pull request:
Using /tensorflow/python/framework/graph_util_impl.py # **convert_variables_to_constants** to freeze graph to pb model will not remove **is_training** flag of **batch_normalization** as well as **fused batch_normalization** ops (tf.layers.batch_normalziation) in the inference graph.  Moreover, the **is_training** flag is still set as true. This issue will not cause any errors to use the pb model on Android. However, it will cause converting errors to tflite model by toco. Is it possible to optimize this conversion tool to resolve the problem in further versions. 

_Following is the node output of the pb model during loading:
name: "convolution_layer/block_layer1/batch_normalization/FusedBatchNorm"
op: "FusedBatchNorm"
input: "convolution_layer/block_layer1/conv2d/Conv2D"
input: "batch_normalization/gamma/read"
input: "batch_normalization/beta/read"
input: "convolution_layer/block_layer1/batch_normalization/Const"
input: "convolution_layer/block_layer1/batch_normalization/Const_1"
attr {
  key: "T"
  value {
    type: DT_FLOAT
  }
}
attr {
  key: "data_format"
  value {
    s: "NHWC"
  }
}
attr {
  key: "epsilon"
  value {
    f: 0.0010000000475
  }
}
**attr {
  key: "is_training"
  value {
    b: true
  }**
}_

powderluv committed:
[XLA] Fix XLA/tfcompile compile error on OSX

This fixes issue #15196

kernel_support_library.cc:99:5: error: no matching function
for call to 'transform' std::transform(function->arg_begin(),..

TEST=build tfcompile on OSX (also requires PR#14893)

tensorflower-gardener committed:
Replace std::pow with MathUtil::IPow when the exponent is a non-negative integer.

Includes minor test cleanup.

PiperOrigin-RevId: 179680628

MarkDaoust committed:
rename files

PiperOrigin-RevId: 179683700

tensorflower-gardener committed:
Add support for constant resolution of the Fill op.

PiperOrigin-RevId: 179684795

Commit:
Fix a flaky test.

In the old code, `a` can be freed before bytes_in_use, making the result of
`bytes_in_use` unstable. In the new code, `b` uses `a` and control-depends on
`bytes_in_use`, so `a` cannot be freed before `bytes_in_use`.

PiperOrigin-RevId: 179690026

allenlavoie committed:
Add some unit tests for memory optimizer _recompute_hint Python usage (mostly
for documentation purposes).

PiperOrigin-RevId: 179697882

Commit:
Support more control flow ops. Delete ops that require ref inputs (e.g.
variables); these ops are not applicable to layout optimizer, because
layout-agnostic nodes are never directly connected to variables, but to
other layout-sensitive nodes such as conv and pooling.

PiperOrigin-RevId: 179699124

rjpower committed:
Reduce the number of warning messages for unsupported operations.

PiperOrigin-RevId: 179700417

tensorflower-gardener committed:
Adding missing comma in __init__.py and fixing links in NOS optimizers docs.

PiperOrigin-RevId: 179701465

tensorflower-gardener committed:
Adding an optimizer module for ConstantSGD.

PiperOrigin-RevId: 179703560

tensorflower-gardener committed:
Adding mention of contrib optimizers and new learning rate decays.

PiperOrigin-RevId: 179706954

tensorflower-gardener committed:
Convert Status-based call sequence checks to CHECK-based.

Call sequence requirements are correctness requirements; any violation of a correctness requirement is a code defect / bug. Currently, some code in the TF GCS client uses Status objects to report call sequence violations. This tolerates bugs, allowing them to go undetected and uncorrected.

This change converts many call sequence checks to use CHECK, rather than Status. The result is code that is simpler, and that exposes bugs, rather than protecting them.

PiperOrigin-RevId: 179709047

facaiy committed:
C++ gradient for Select (#14862)

* ENH: add SelectGrad

* TST: add test case

* TST: clean code

* CLN: grad_scope -> scope

* CLN: x_shape, y_shape -> shape

yifeif committed:
Merge pull request #15524 from uber/fix-golang-readme-formatting

Fix Golang readme installation instruction formatting

yifeif committed:
Merge pull request #15519 from ManHyuk/fix_typo

fix typos

yifeif committed:
Merge pull request #15504 from snnn/f901

Fix typos

kagemusha committed:
Update debugger.md

yifeif committed:
Merge pull request #15478 from rongjiecomputer/env

[XLA/tfcompile] Add Env::CreateUniqueFileName and use it in SaveGraph

rjpower committed:
Fix logging of excessive unsupported operations.

An operator precedence issue with the previous version triggered errors when models had a large number of summaries.

PiperOrigin-RevId: 179722697

yifeif committed:
Merge pull request #15229 from powderluv/fix_osx_tfcompile

[XLA] Fix another XLA/tfcompile compile error on OSX

cdleary committed:
[XLA] Remove inappropriate plugin linkage.

PiperOrigin-RevId: 179724373

yifeif committed:
Add no_pip tag to scan_dataset_op_test

Commit:
Support SplitV op.

PiperOrigin-RevId: 179725704

hawkinsp committed:
Restore escaping to string attribute list summaries accidentally removed by previous change.

PiperOrigin-RevId: 179725715

saeta committed:
Add prefetching into parallel_interleave

This change adds 2 parameters to parallel_interleave:
 - prefetch_input_elements: determines the number of iterators to prefetch
     allowing buffers to warm up and data to be pre-fetched without blocking
     the main thread (i.e. the GetNext() call).
 - buffer_output_elements: in order to avoid creating thousands of threads, we
     fuse in the .prefetch() operator as an additional parameter. The value of
     this parameter is identical to the value passed to `.prefetch()`

PiperOrigin-RevId: 179726088

Commit:
Extend DataFormatDimMap to handle tensors.

PiperOrigin-RevId: 179726269

tensorflower-gardener committed:
Destroy batch_thread_pool_ first. For the new in flight batches implementation, the callbacks scheduled on this thread pool refer to other class members which must remain alive until the thread pool is empty.

Also fix a similar lifetime issue in the unit tests.

PiperOrigin-RevId: 179726389

yifeif committed:
Merge pull request #15507 from rongjiecomputer/path

[XLA] Use os.path for path manipulation

angerson committed:
Make tf_cc_test compilable by default for Android

This change adjusts the build options for tf_cc_test targets so that
they can be built for Android devices by default. Many targets will
still need additional options on a per-target basis.

Android requires that executables be compiled with `-pie`.

PiperOrigin-RevId: 179729092

tensorflower-gardener committed:
Add the capability of verbose logging for HTTP requests to the TensorFlow GCS client.

This change allows the GCS client within TensorFlow to log verbose information about HTTP requests to the standard Google logger.  (Actually, to the TensorFlow variant of the standard Google logger.)

This capability is disabled by default. It is enabled by setting the environment variable GCS_LOG_HTTP_REQUEST_VERBOSE=1 (or =true).

PiperOrigin-RevId: 179729641

yifeif committed:
Merge branch 'master' into master

yifeif committed:
Merge pull request #14673 from GreggHelt2/master

Fixed bug in code within programmer's guide markdown docs for Variabl…

tensorflower-gardener committed:
Adds a randomized test for Atan.
Resolves TODOs that are out of scope because conv op registration is non-C64 only.

PiperOrigin-RevId: 179730049

Commit:
[XLA] Add reassociation rule for adds, and add canonicalization rules for add/sub of a constant.

This patch adds a new algebraic simplification:

* (A + C0) + C1 => A + (C0 + C1), where C0 and C1 are constants.  This
  allows us to constant-fold C0 + C1.

In service of this rule, this patch also adds two new canonicalizations:

* Const + A => A + Const
* A - Const => A + (-1 * Const)

PiperOrigin-RevId: 179731747

yifeif committed:
Merge pull request #14332 from pvaneck/jcenter-android

Update Android JCenter links

hawkinsp committed:
[TF:XLA] Move XlaResource into its own file, and refactor it into a better-abstracted class. No functional changes intended.

PiperOrigin-RevId: 179734920

alextp committed:
Implements the fill op for int64 indices.

PiperOrigin-RevId: 179735577

saeta committed:
Truncate status messages before converting to gRPC.

PiperOrigin-RevId: 179736746

yifeif committed:
Merge pull request #15260 from ManHyuk/typo

Fix typo

yifeif committed:
Merge pull request #14527 from wodesuck/patch-1

Fix control input name for quantize node

yifeif committed:
Merge pull request #14478 from qmick/slice_op_remove_redundant

Remove redundant code of slice_op.cc

annarev committed:
Automated g4 rollback of changelist 179729641

PiperOrigin-RevId: 179742670

yifeif committed:
Merge pull request #15516 from rongjiecomputer/cxx

[XLA] Define LANG_CXX11 for >= VS 2015

yifeif committed:
Merge pull request #14377 from footh/layer-base-indent-fix

Fix indenting on cache check in Network layer class

yifeif committed:
Merge pull request #14403 from ZhengshengWei/r-2017-11-09

modified convolution document

yifeif committed:
Merge pull request #14409 from tano297/jetson_platform

added patch needed for jetson tx2/1

yifeif committed:
Merge pull request #15512 from rongjiecomputer/hide_gcc_workaround

[XLA] Hide GCC 7.1.1 workaround from MSVC

yifeif committed:
Merge pull request #15511 from rongjiecomputer/array

[XLA] Fix std::array initialization

yifeif committed:
Merge pull request #15509 from rongjiecomputer/numeric

[XLA] Explicitly include <numeric>

yifeif committed:
Merge pull request #15508 from rongjiecomputer/const

[XLA] Add 'const' to custom comparator

Commit:
Support max pool v2.

PiperOrigin-RevId: 179747281

rongjiecomputer committed:
[XLA] Define ssize_t for Windows

yifeif committed:
Merge pull request #14259 from vnavkal/svd

Implement backpropagation for SVD when full_matrices is False

alextp committed:
implicit_gradients always sorts its return values by variable creation order

PiperOrigin-RevId: 179750392

rongjiecomputer committed:
Remove old macro implementation

froystig committed:
[XLA] Add Collapse and Rev ops to the local XLA Python client.

PiperOrigin-RevId: 179753056

rongjiecomputer committed:
Update comment

Commit:
Fix the condition check for int8 fused_conv so that it works on Volta.

PiperOrigin-RevId: 179756490

tensorflower-gardener committed:
Fix typo in struct name: RecvNodeDescritorHash -> RecvNodeDescriptorHash.

PiperOrigin-RevId: 179756840

Commit:
[XLA] Fix HLO graph dumper not to assume that instruction names start with "%".

HLO graph dumper needs to be aware that we've gotten rid of the "%"
prefix in HLO names so it doesn't print e.g.

  reduce
  reduce.42
  Subcomputation: add
  ...

but instead simply prints

  reduce.42
  Subcomputation: add
  ...

PiperOrigin-RevId: 179756922

cdleary committed:
[XLA] Add CrossReplicaSum to Python API bindings.

Add support to CPU for the trivial replica_count=1 case.

PiperOrigin-RevId: 179756956

tensorflower-gardener committed:
Update ops-related pbtxt files.

PiperOrigin-RevId: 179757188

caisq committed:
Merge pull request #15528 from kagemusha/patch-1

Update debugger.md

yifeif committed:
Change sys.maxint to -1 in hook.py

yifeif committed:
update sys.maxint to 1 in grpc_wrapper.py

LucasSloan committed:
Use a trailing period to define floating point constant.

In the code for tf.estimator, the array y_eval uses the constant "-7", when the rest of surrounding constants use trailing periods.  The corresponding code for a custom estimator has the trailing period.

annarev committed:
Automated g4 rollback of changelist 179735577

PiperOrigin-RevId: 179761787

tensorflower-gardener committed:
Go: Update generated wrapper functions for TensorFlow ops.

PiperOrigin-RevId: 179761889

yifeif committed:
Merge pull request #15536 from LucasSloan/fix-get-started-docs

Use a trailing period to define floating point constant.

Commit:
Remove quantized_matmul_op_for_hexagon_test in Windows build scripts

yifeif committed:
Merge pull request #15526 from ekelsen/branch_179628764

Branch 179628764

annarev committed:
Automated g4 rollback of changelist 179757188

PiperOrigin-RevId: 179766237

mrry committed:
[tf.data] Add `tf.contrib.data.parse_single_example()`.

The new op is a fused implementation of the existing
`tf.parse_single_example()`, which is more efficient when parsing a
single Example at a time.

PiperOrigin-RevId: 179768512

tensorflower-gardener committed:
Update ops-related pbtxt files.

PiperOrigin-RevId: 179769207

annarev committed:
Internal change.

PiperOrigin-RevId: 179769382

tensorflower-gardener committed:
Go: Update generated wrapper functions for TensorFlow ops.

PiperOrigin-RevId: 179769852

facaiy committed:
TST: unknown shape case

facaiy committed:
BUG: fix for unknown shape

Commit:
Updates imageops Docs

mrry committed:
Revert change to ParseBytesList that failed on some platforms.

PiperOrigin-RevId: 179775354

ebrevdo committed:
[tf nest] Add additional key yielder.

This is just copying a utility function created by Malcolm Reynolds.

PiperOrigin-RevId: 179775504

Commit:
fix random_distributions_test

tensorflower-gardener committed:
Roll CL 177989542 forward with fix: Wrappers for CUDA 9 warp-synchronous intrinsics.

PiperOrigin-RevId: 179782067

MarkDaoust committed:
Modernize old "get_started/get_started.md", as "programmers_guide/low_level_intro.md".

PiperOrigin-RevId: 179807033

tensorflower-gardener committed:
Implement Clone for kSend,kRecv,kSendDone and kRecvDone.

PiperOrigin-RevId: 179808928

anight committed:
golang: added Session.ListDevices method (#14385)

* golang: added Session.ListDevices method

theflofly committed:
Merge with master.

caisq committed:
tfdbg: allow writing tensor values to filesystem as numpy files

Command example:
tfdbg> pt Softmax:0 -w /tmp/softmax.npy
tfdbg> eval 'np.sum(`Softmax:0`, axis=1)' -w /tmp/softmax_sum.npy

The saved file can later be loaded in Python:
>> import numpy as np
>> np.load('/tmp/softmax.npy')
>> np.load('/tmp/softmax_sum.npy')

RELNOTES: For `pt` and `eval` commands, allow writing tensor values to filesystem as numpy files.
PiperOrigin-RevId: 179817453

langmore committed:
sparse_ops:  Static shape update.

* .sparse_reshape:  Set static shape on output when shape includes inferred dims.
* .transpose:  Set static shape if all shapes are known.

PiperOrigin-RevId: 179820175

ebrevdo committed:
[TFE] Return a byte-encoded value for tfe.defun's name

This is necessary when the function object is passed to an op constructor;
the op's func attribute is a protobuf field and must be bytes in python3.

PiperOrigin-RevId: 179822007

yifeif committed:
Merge pull request #15539 from snnn/fix1

Remove quantized_matmul_op_for_hexagon_test in Windows build scripts

yifeif committed:
Merge pull request #15515 from rongjiecomputer/macro

[XLA] Use simplified version of TF_ASSIGN_OR_RETURN for MSVC

yifeif committed:
Merge pull request #15510 from rongjiecomputer/aligned_mem

[XLA] Use tensorflow::port::Aligned{Malloc,Free}

yifeif committed:
Merge pull request #15436 from tjingrant/sru

Simple Recurrent Unit

yifeif committed:
Merge pull request #15506 from snnn/fix10

bugfix: long is 32 bits on Windows

yifeif committed:
Merge commit for internal changes

yifeif committed:
Merge pull request #15550 from snnn/random_distributions_test

fix random_distributions_test

Commit:
Adds eval session master into TF_CONFIG

PiperOrigin-RevId: 179835208

tensorflower-gardener committed:
[XLA] Make dumped_computation_to_text print the short parsable hlo text.

PiperOrigin-RevId: 179835513

yifeif committed:
Merge pull request #15513 from rongjiecomputer/ssize

[XLA] Define ssize_t for Windows

yifeif committed:
Merge pull request #15544 from facaiy/BUG/unknown_shape_for_sparse_multiclass_hing_loss

support unknown shape for `sparse_multiclass_hing_loss`

yifeif committed:
Merge pull request #14422 from theflofly/add-gradient-prod-op

Add the C++ gradient of the Prod operation.

tensorflower-gardener committed:
Adding support for to resolve constant FloorDiv, FloorMod, StridedSlice, Stack, Rank and Range.

PiperOrigin-RevId: 179836027

Commit:
Fix padding for int8 fused convolution.

cudnnConvolutionBiasActivationForward doesn't work when filter size >= 6 and
there is padding. This CL works around this cudnn bug.

PiperOrigin-RevId: 179836819

akshayka committed:
This change adds a mechanism to the internal C API for updating an output handle's shapes and types after its source operation has been created.

Context: framework/ops.py was recently updated to use the C API when setting shapes for an op's outputs. This update broke shape inference for graph functions that captured resource handles; this, in turn, made it impossible to create graph functions from Python methods that required fully defined shapes (e.g., like MNIST's `call` method). In particular, the C API computes shapes for ops when they are created and does not update them thereafter; this is problematic because when a resource handle is captured while building a function, we need to update the captured operation's output handle in order to propagate its outputs shapes and dtypes.
PiperOrigin-RevId: 179837104

ManHyuk committed:
fix typo

ManHyuk committed:
fix typo

MarkDaoust committed:
Add links to low level API intro

PiperOrigin-RevId: 179844300

benoitsteiner committed:
Tagged the grappler python tests

PiperOrigin-RevId: 179844728

yifeif committed:
Merge pull request #15562 from ManHyuk/fix_typo

Fix typos

yifeif committed:
Merge pull request #15545 from wagonhelm/updatedocs

Update Imageops Docs

yifeif committed:
Merge pull request #15560 from yifeif/branch_179822007

Branch 179822007

akshayka committed:
Make it possible to wrap tf.layers.Layer's 'call' method in tfe.defun.

This change wraps Layer's `build` method in an `init_scope`, which has the effect of lifting all ops created in `build` out of function-building graphs and control flow scopes; the `init_scope` also pauses the gradient tape.

PiperOrigin-RevId: 179846085

gunan committed:
Automated g4 rollback of changelist 179836027

PiperOrigin-RevId: 179848421

tensorflower-gardener committed:
Merged commit includes the following changes:
179858783  by gunan:

    [XLA] Add a pass to replace all non-side-effecting and removable zero sized
    HLOs.

--
179853916  by A. Unique TensorFlower:

    Fix missing backtick in parse_example docstring, causing wrong markdown formatting.

--
179852823  by apassos:

    Another attempt at adding int64 indices to Fill.

--
179852553  by A. Unique TensorFlower:

    Stop grappler optimization to reorder BiasAdd ops.

--
179850703  by yaozhang:

    Support reverse op.

--
179850204  by blakehechtman:

    [XLA] Add a pass to replace all non-side-effecting and removable zero sized
    HLOs.

--

PiperOrigin-RevId: 179858783

yifeif committed:
Merge pull request #14785 from shreyneil/master

Update layers pull request

tensorflower-gardener committed:
Automated g4 rollback of changelist 179782067

PiperOrigin-RevId: 179861781

jsimsa committed:
Saving a couple of us in IteratorGetNextOp::ComputeAsync.

PiperOrigin-RevId: 179863871

tensorflower-gardener committed:
Only squeeze the last dimension of outputs and indices in infer_graph of
kmeans.

PiperOrigin-RevId: 179865588

annarev committed:
Internal change.

PiperOrigin-RevId: 179866603

Commit:
[XLA] Transform division-by-fp-constant to multiplication by its inverse.

Backends can do this transformation, but in general they won't if the
divisor is not a constant.  In addition, this formulation allows more
aggressive fusion.

PiperOrigin-RevId: 179869591

blakehechtman committed:
Automated g4 rollback of changelist 179858783

PiperOrigin-RevId: 179870508

tensorflower-gardener committed:
Update ops-related pbtxt files.

PiperOrigin-RevId: 179872583

benoitsteiner committed:
Extract placement restrictions for a graph

PiperOrigin-RevId: 179872807

Commit:
Fold FuseBatchNorm correctly.

PiperOrigin-RevId: 179874901

mrry committed:
Add test coverage for example parsing with a feature that is an empty string.

PiperOrigin-RevId: 179875758

benoitsteiner committed:
Added a comment to explain why we can't reorder BiasAdd

PiperOrigin-RevId: 179878687

jsimsa committed:
Adding a test that checks that Grappler is able to infer shapes for core `tf.data` API.

PiperOrigin-RevId: 179878973

annarev committed:
Internal change.

PiperOrigin-RevId: 179880234

Commit:
No external change.

PiperOrigin-RevId: 179881638

annarev committed:
Automated g4 rollback of changelist 179872807

PiperOrigin-RevId: 179882393

Commit:
[XLA:GPU] Implement BatchNormThunk as a call into cudnn.

Using cudnn for these calls is disabled by default, because it's not a
performance win on our benchmarks.

PiperOrigin-RevId: 179882911

Commit:
[XLA] Run ZeroSizedHloElimination after BatchNormExpander.

BatchNormExpander can create zero-sized HLOs even if it's run after
ZeroSizedHloElimination.  For example, you might have a
BatchNormTraining op with shape (f32[0, 10], f32[10], f32[10]).  This
won't be eliminated by ZeroSizedHloElimination, but when it's
decomposed, it will result in many zero-sized HLOs.

PiperOrigin-RevId: 179885417

Commit:
Support fill op.

PiperOrigin-RevId: 179888184

Commit:
[XLA] Add zero-sized batchnorm test.

Also fix out-of-bounds read in ReferenceUtil::Reduce4DTo1D.

PiperOrigin-RevId: 179888724

tensorflower-gardener committed:
[XLA] Update the parsing and stringfication of the Dot dimension numbers.

PiperOrigin-RevId: 179888764

Commit:
[XLA] Print out missing extra-info for many instructions in the HLO graph dumper.

Now we use the same functionality as HloInstruction::ToString() to print
instructions' extra info.  This fills in a lot of previously-missing
info, like reduce-windows' windows, and dots' dot-dimension-numbers.

PiperOrigin-RevId: 179892469

benoitsteiner committed:
Made hard colocation constraints (i.e constraints that must be met for the
model to be executable) available from python

PiperOrigin-RevId: 179892785

isaprykin committed:
Support multiple loss and multiple optimizers in replicate_model_fn.

Instead of supplying `optimizer_fn`, the user is now expected to wrap their optimizer in GatheringOptimizer.  The latter will gather gradients, reduce and apply them.

There can be multiple instances of GatheringOptimizer inside the model.

PiperOrigin-RevId: 179899422

tensorflower-gardener committed:
[XLA] Support conditional in all backends.

PiperOrigin-RevId: 179900775

tensorflower-gardener committed:
Added an option to download a fresh release of clang for doing cuda_clang builds.

Tested only on Linux, but should also work for Mac.
No support for Windows is available yet.

PiperOrigin-RevId: 179910980

tensorflower-gardener committed:
Add comments to link convolutional factors code with definitions in the KFC paper.

PiperOrigin-RevId: 179925679

mrry committed:
Mark the Arg and RetVal ops as inexpensive.

This change avoids gratuitous context switches at the beginning of a
step (for each fed value) and end of a step (for each fetched
value). These OpKernels are trivial, so the cost of executing it
inline is typically smaller than waking up another thread to perform
the work.

PiperOrigin-RevId: 179933412

Commit:
No need to and don't add layout transform to control nodes.

PiperOrigin-RevId: 179934839

Commit:
[XLA:GPU] Fix implementation of Thunk::ShouldBlockFutureThunks.

The old implementation did not do what it was intended to do.  Instead
of making the next thunk on each other stream wait for this thunk to
complete, it made all following thunks on this stream wait for this
thunk to complete.  (Which is redundant, since streams are serialized
anyway.)

PiperOrigin-RevId: 179938424

benoitsteiner committed:
Extract placement restrictions for a graph

PiperOrigin-RevId: 179939679

tensorflower-gardener committed:
Adds support in stream executor interface to update the scratch allocator used with a cuFFT plan. This enables plan reuse without requiring we keep the scratch allocation alive between executions.

PiperOrigin-RevId: 179939994

Commit:
Fix allow_smaller_final_batches for bucket_by_sequence_length.
Closes #14420.

PiperOrigin-RevId: 179940036

tensorflower-gardener committed:
In FirstReady node manager, use node name as tie-breaker when multiple nodes
have same time_ready.

PiperOrigin-RevId: 179940344

hawkinsp committed:
[XLA] Remove Power architecture support from CPU backend.

The Power support is unused, untested, and buggy. Remove it until it is fixed.

PiperOrigin-RevId: 179942340

jsimsa committed:
Fixing tf.contrib.data API declarations.

PiperOrigin-RevId: 179943582

hawkinsp committed:
[TF:XLA] Implement AdjustContrastv2, AdjustHue, AdjustSaturation, HSVToRGB, RGBToHSV in XLA bridge.

PiperOrigin-RevId: 179945310

langmore committed:
assert_near added to check_ops.py

PiperOrigin-RevId: 179946066

jsimsa committed:
Adding performance guide for the `tf.data` API.

PiperOrigin-RevId: 179947300

alextp committed:
Uses ImportGraphDef instead of ConvertGraphDefToGraph in tflite.

This has the right semantics for backwards compatibility.

PiperOrigin-RevId: 179951639

benoitsteiner committed:
Updated the virtual cluster to return the proper error code if the simulated
peak memory usage exceeds the available memory.

PiperOrigin-RevId: 179952918

tensorflower-gardener committed:
Merge changes from github.

PiperOrigin-RevId: 179953488

tensorflower-gardener committed:
Go: Update generated wrapper functions for TensorFlow ops.

PiperOrigin-RevId: 179954704

claynerobison committed:
Adding MKL-DNN support for LRN op

Commit:
merge

tensorflower-gardener committed:
Update ops-related pbtxt files.

PiperOrigin-RevId: 179955796

tensorflower-gardener committed:
Adds more useful debug information to an error message when no kernels are found.

PiperOrigin-RevId: 179957977

yifeif committed:
Remove duplicate definition of compression_type

caisq committed:
[TFE] Add link to other exapmles from spinn/README.md

PiperOrigin-RevId: 179960166

mrry committed:
Fix typo in direct_session_test.cc.

PiperOrigin-RevId: 179960797

Commit:
Support non-constant param input of AvgPoolGrad and Sum.

PiperOrigin-RevId: 179962212

tensorflower-gardener committed:
Fix documentation error in tensorflow feature_utils.h

Need to take address of the example for modifying its feature

PiperOrigin-RevId: 179963317

tensorflower-gardener committed:
Automated g4 rollback of changelist 179846085

PiperOrigin-RevId: 179963793

langmore committed:
Use assert_near in mvn_full_covariance to assert symmetric.

PiperOrigin-RevId: 179964717

Commit:
Extracts the tpu_job_name from TF_CONFIG if set.

PiperOrigin-RevId: 179966184

tensorflower-gardener committed:
Make the analytical estimator and the virtual cluster own the ready_node_manager to be used   by the virtual_scheduler. This allows us to customize the ready_node_manager for each analyzer instance.

PiperOrigin-RevId: 179967778

tensorflower-gardener committed:
Output of a slice op can alias its operand.

PiperOrigin-RevId: 179969317

tensorflower-gardener committed:
[XLA] Hlo parser: update readme.

PiperOrigin-RevId: 179969357

tensorflower-gardener committed:
Adding page to tensorflow.org with directions for building the TFLite demo on Android.

PiperOrigin-RevId: 179970218

mrry committed:
Avoid unnecessary copying and allocation in `ExecutorBarrier::WhenDone()`.

Previously the code would make a copy of the `done_cb_` and update the status
on a successful completion, neither of which is necessary, and both of which
showed up in trace amounts in profiling.

PiperOrigin-RevId: 179970701

Commit:
Run gen_git_source.py inside of a repo_rule instead of configure.

PiperOrigin-RevId: 179971055

tensorflower-gardener committed:
Replaces custom _lengths_to_masks function with the official, more efficient sequence_mask function that supersedes it.

PiperOrigin-RevId: 179971521

yifeif committed:
Merge pull request #15593 from ekelsen/branch_179953488

Branch 179953488

benoitsteiner committed:
Shard and turn on the layout optimizer test
Tag the dataset test

PiperOrigin-RevId: 179974079

alextp committed:
Allow ImportGraphDef to ignore shape validation to match semantics of ConvertGraphDefToGraph optionally.

PiperOrigin-RevId: 179974220

alextp committed:
Produces RunMetadata from eager execution when tracing is explicitly enabled.

This will allow downstream profiling tools to work with eager execution.

PiperOrigin-RevId: 179974238

jsimsa committed:
Changing tf.contrib.data.dense_to_sparse_batch to return tf.SparseTensor.

PiperOrigin-RevId: 179975122

sguada committed:
Clarify Eager Metrics docs.
Add Metric.value()

PiperOrigin-RevId: 179976743

tensorflower-gardener committed:
Update ops-related pbtxt files.

PiperOrigin-RevId: 179976854

tensorflower-gardener committed:
Go: Update generated wrapper functions for TensorFlow ops.

PiperOrigin-RevId: 179977826

tensorflower-gardener committed:
Remove a few unused constructions and simplfy some code

PiperOrigin-RevId: 179978470

Commit:
Adds train_hooks into _TrainingExecutor.

PiperOrigin-RevId: 179978502

Commit:
Fill the 4th and 5th output of FusedBatchNormGradOp with zeros so that check_op won't fail.

PiperOrigin-RevId: 179981448

tensorflower-gardener committed:
Adds FFT for XLA: CPU via Eigen, GPU via cuFFT.

GPU support includes plan reuse with new scratch allocator per execution in fft_thunk.

PiperOrigin-RevId: 179983419

tensorflow-jenkins commented:
Can one of the admins verify this patch?

tensorflower-gardener committed:
Automated g4 rollback of changelist 179983419

PiperOrigin-RevId: 180000981

hawkinsp committed:
[TF:XLA] Refactor large list of hard-coded compile-time constant arguments to operators. Add a new .CompileTimeConstInput() annotation on kernel registrations instead.

PiperOrigin-RevId: 180008567

terrytangyuan committed:
Update tf-learn reference to TF Estimators (#15596)

caisq committed:
Merge pull request #15594 from Intel-tensorflow/pr-mkl-dnn-lrn

MKL: Adding MKL-DNN support for LRN op

caisq committed:
Merge commit for internal changes

protoget committed:
Assign subgraph created by cudnn_rnn saveable to the 1st GPU.

PiperOrigin-RevId: 180053468

drpngx commented:
Can you pull rebase and push again? It's hard to tell what you're trying to merge.

caisq committed:
Merge pull request #15615 from caisq/branch_180008567

Branch 180008567

him2him2 committed:
Update: HTTP -> HTTPS

URLs updated to use HTTPS protocol where appropriate to improve security and privacy.

claynerobison committed:
Reverting PR #14478, which breaks Inception v3 with MKL. (#15566)

Commit:
Merge commit for internal changes

Androbin committed:
Clean bazel `all_files` targets (#15368)

* Clean bazel `all_files` targets

* Fix sanity tests

facaiy committed:
add c++ gradient for op: Pow (#15245)

* TST: add test case
* ENH: add pow grad

caisq committed:
Merge pull request #15632 from drpngx/branch_180053468

Branch 180053468

caisq committed:
Merge pull request #15629 from him2him2/fix/HTTPS

Update: HTTP -> HTTPS

drpngx commented:
Could you pull rebase and push again?

martinwicke committed:
Remove unwrapping for function argument inspection in Estimator.

PiperOrigin-RevId: 180147476

nluehr committed:
Add support for CUBLAS_TENSOR_OP_MATH in fp16 GEMM (#13451)

- Applies to matrix multiplications with fp16 input/output.
  Computations will fall back to pseudo-fp16 if tensor op math is
  disabled or not supported.
- Enabled by default. Tensor ops (both in cublas gemms and cudnn
  convolutions) can be disabled globally by setting the
  environment variable TF_DISABLE_TENSOR_OP_MATH=1. To disable
  tensor ops specifically for gemms or convolutions use
  TF_DISABLE_CUBLAS_TENSOR_OP_MATH=1 or
  TF_DISABLE_CUDNN_TENSOR_OP_MATH=1, respectively.
- Added CUBLAS 9.0 algorithms to GetBlasGemmAlgorithms().

yaroslavvb committed:
Remove unneeded branch check (#13495)

* Remove branch check

* Change version to read “inconsitent-git-version”

claynerobison committed:
Fixing MKL-DNN convolution filter propagation to backprop (#15227)

shahidhs-ibm committed:
Update C API test data comparison for s390x (#15390)

* Correct the AppendHash test data comparison for s390x

* Correct _USE_C_API test data comparison for s390x

* Correcting the Python coding style (pylint)

Correcting the following comment issued by pylint:
C: 25, 0: standard import "import platform" should be placed before "import numpy as np" (wrong-import-order)

* Correcting code to conform to Google C++ Style Guide

Correcting as per the changes suggested by "clang-format".

* Making the changes generic for all big endian architectures

Changes made as per review comment: https://github.com/tensorflow/tensorflow/pull/15390#pullrequestreview-84306138

* Making the changes generic for all big endian architectures

Changes made as per review comment: https://github.com/tensorflow/tensorflow/pull/15390#discussion_r157631673

* Correcting code to conform to Google C++ Style Guide

nhasabni committed:
Adding missing reorders in ReLU and AddN (#15406)

yongtang committed:
Add shape function SingleImageRandomDotStereograms (#15431)

* Add shape function SingleImageRandomDotStereograms

This fix tries to address the issue raised in 15429 where
there is no shape function for SingleImageRandomDotStereograms.

This fix adds the shape function for `SingleImageRandomDotStereograms`.

NOTE: `SingleImageRandomDotStereograms` takes an attribute
of `output_image_shape` which is in the format of
`[X, Y, C]` (`[ImageX, ImageY, Channel]`. However, the actual
data output is in the format of
`[ImageY, ImageX, Channel]` (`[h, w, c]`)
so by default the output_image_shape has the value
of [1024, 768, 1] but the output data will be [768, 1024, 1].
And if `[1200, 800, 1]` is used explicitly then the output
data shape will be `[800, 1200, 1]`.

This fix does not change the behavior for now.

This fix fixes 15429.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

* Add test cases for shape function for SingleImageRandomDotStereograms

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

* Update Bazel BUILD file

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

* Address review feedback.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

errord committed:
call within the loop (#15446)

* call within the loop

optimize call within the loop

* const int

Commit:
Add ws2_32.lib to gcs_dns_cache_test's linkopts (#15496)

rongjiecomputer committed:
[XLA] Remove unused dlfcn.h and implement sincos[f] for MSVC (#15514)

* [XLA] Remove unused dlfcn.h and implement sincos[f] for MSVC

* Move sincos[f] to windows_compatibility.h

tensorflow/compiler/xla/service/cpu/simple_orc_jit.cc
tensorflow/compiler/xla/service/cpu/windows_compatibility.h

* Move implementation to windows_compatibility.cc

* Sort srcs

* Add comment

ngc92 committed:
added note about weights gradient in compute_weighted_loss (#15533)

Commit:
Bug fix for example_parsing_ops_test (#15548)

Static variables' initialization order is not determined in C++.
In a static variable's constructor, you can't access other static
variables unless its constructor is a constexpr, which is not true
for tensor's allocators.

Replacing it with std::call_once.

yanboliang committed:
Fix adding elements to collections.deque. (#15574)

rongjiecomputer committed:
[Bazel/MSVC] Fix build error since aae439c (#15579)

* [Bazel/MSVC] Fix build error since aae439c

* Revert "[Bazel/MSVC] Fix build error since aae439c"

This reverts commit 635f6202f312165b5e29c2ef9cf70f1614b4043e.

* [Bazel/MSVC] Fix build error since aae439c

EncodeTS committed:
fix InvalidArgumentError when using cv2 with tf.py_func() (#14286)

caisq committed:
Merge pull request #14999 from n3011/patch-1

fix clip weights tests

drpngx committed:
Remove unwrapping for function argument inspection in Estimator. (#15646)

PiperOrigin-RevId: 180147476

yupbank committed:
consistency in function name.. (#14260)

yongtang committed:
Add templated functions for `safe_strto[f|d|i32|i64|u32|u64]` (#14789)

* Add templated functions for `safe_strto[f|d|i32|i64|u32|u64]`

While working on 14330 I noticed that there is no templated functions for
`safe_strto[f|d|i32|i64|u32|u64]`. As a result, an additional wrapper
has to be created in different places to apply the typename in a
templated class or function. Examples include the existing implementations
in `tensorflow/core/kernels/string_to_number_op.cc` (`Convert`),
`tensorflow/core/lib/strings/proto_text_util.h` (`ProtoParseNumeric`),
and in 14330.
It might make sense to add a templated function for `safe_strto[f|d|i32|i64|u32|u64]`
to avoid existing and future code duplications.
This fix adds
```
template<typename T>
bool SafeStringToNumeric(StringPiece s, T* value);
```
to address the above mentioned issue.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

* Move ProtoParseNumeric to numbers.h, and use overloading in template

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Add int64 support for BroadcastArgs and BroadcastGradientArgs (#15255)

* Add int64 support for BroadcastArgs and BroadcastGradientArgs

In `array_ops.cc`, both int32 and int64 are expected to be supported
for `BroadcastArgs` and `BroadcastGradientArgs`. However, this was
not the case as only int32 kernel are registered even though `T`
is part of the `TypeConstraint`.

This fix adds the int64 kernel support for `BroadcastArgs` and
`BroadcastGradientArgs`, and adds related test cases.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

* Add test cases for int64 support of BroadcastArgs and BroadcastGradientArgs

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

* Sanitize python and clang-format.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

DavidNorman committed:
Make the client_test able to be disabled using a manifest file (#15277)

yongtang committed:
Fix discrepancy between doc and impl for `tf.reverse` (#13672)

* Fix discrepancy between doc and impl for `tf.reverse`

This fix tries to address the discrepancy between doc and
implementations for `tf.reverse`. In the documentation,
both int32 and int64 could be used for axis.

However, the actual implementation of the `tf.reverse` does
not support int64 and caused missing kernel error:
```python
ubuntu@ubuntu:~$ python
Python 2.7.12 (default, Nov 19 2016, 06:48:10)
[GCC 5.4.0 20160609] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import tensorflow as tf
>>> v = tf.reverse_v2([1, 2, 3], tf.constant([0], tf.int64))
>>> sess = tf.Session()
>>> sess.run(v)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/client/session.py", line 889, in run
    run_metadata_ptr)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/client/session.py", line 1120, in _run
    feed_dict_tensor, options, run_metadata)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/client/session.py", line 1317, in _do_run
    options, run_metadata)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/client/session.py", line 1336, in _do_call
    raise type(e)(node_def, op, message)
tensorflow.python.framework.errors_impl.InvalidArgumentError: No OpKernel was registered to support Op 'ReverseV2' with these attrs.  Registered devices: [CPU], Registered kernels:
  device='CPU'; T in [DT_STRING]; Tidx in [DT_INT32]
  device='CPU'; T in [DT_BOOL]; Tidx in [DT_INT32]
  device='CPU'; T in [DT_COMPLEX128]; Tidx in [DT_INT32]
  device='CPU'; T in [DT_COMPLEX64]; Tidx in [DT_INT32]
  device='CPU'; T in [DT_DOUBLE]; Tidx in [DT_INT32]
  device='CPU'; T in [DT_FLOAT]; Tidx in [DT_INT32]
  device='CPU'; T in [DT_HALF]; Tidx in [DT_INT32]
  device='CPU'; T in [DT_INT8]; Tidx in [DT_INT32]
  device='CPU'; T in [DT_UINT8]; Tidx in [DT_INT32]
  device='CPU'; T in [DT_INT16]; Tidx in [DT_INT32]
  device='CPU'; T in [DT_UINT16]; Tidx in [DT_INT32]
  device='CPU'; T in [DT_INT32]; Tidx in [DT_INT32]
  device='CPU'; T in [DT_INT64]; Tidx in [DT_INT32]

         [[Node: ReverseV2 = ReverseV2[T=DT_INT32, Tidx=DT_INT64](ReverseV2/tensor, Const)]]

```
This fix addresses this issue and added the int64 support
for axis in `tf.reverse`.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

* Add GPU support of int64 axis in `tf.reverse`

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

* Add test cases for int64 support of axis in `tf.reverse`

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

* Add OpenCL int64 support of axis in `tf.reverse`

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

facaiy committed:
New metric: Cohen's kappa (#15443)

* ENH: implement cohen_kappa

* TST: add test cases

* DOC: add document

* DOC: revise document as suggested

* CLN: drop total prefix

* CLN: wrap lines to 80 chars in metric_ops.py

* CLN: remove unneeded dependency

* DOC: support labels datatype

* TST: add explanation for testBasic

* TST: remove testBasic2

* TST: replace all random data

* CLN: remove sklearn dependency

* TST: wrap the lines in this test to 80 chars

* DOC: clean format

* CLN: typo

* CLN: lables specifiy -1 axis

* CLN: remove float lables support

* DOC: modify document of _safe_div

* CLN: rename predictions -> predictions_idx

* TST: invalid dim case

* CLN: clean codes

* DOC: remove total_ in update_op

jthestness committed:
mpi_collectives: Refactor to fix build issues (#15534)

* mpi_collectives: Refactor to fix build issues

After TF commit 5c7f9e3, the mpi_collectives package would no longer build
ops and kernels. This build issue caused mpi_collectives import to fail in
Python with the following error: "NameError: Could not find operator MPISize
in dynamic library mpi_collectives.so".

To fix this issue, add build targets to ensure both ops and kernels are built
Note, also refactored the build targets and directory structure to more
closely match other contrib packages.

* mpi_collectives: Minor BUILD fix

* Minor: Buildifier fix

* mpi_collectives: Correct preprocessor defines

* mpi_collectives: Clearer defines inclusion

facaiy committed:
Recover the lost output channel number of atrous convolution for NC format (#14120)

* TST: add test case

* BUG: fix unknown spatial dim for NC format

* Make comment more informative

Commit:
session_clusterspec_prop_test and common_runtime_direct_session_with_tracking_alloc_test fixed for GPU (#14530)

* session_clusterspec_prop_test fix for GPU

* common_runtime_direct_session_with_tracking_alloc_test fix for GPU

caisq committed:
tfdbg: Fix a bug in source file content reading

Previously, calling strip() removed not only the trailing newline character, but also any leading spaces and tabs, which actually should be preserved. This CL fixes it.

PiperOrigin-RevId: 180198179

tensorflower-gardener committed:
Optionally store the status code/message in the response
body for RunGraph and RunStep RPCs, to workaround the fact that the
RPC subsystem truncates long metadata messages.

PiperOrigin-RevId: 180203356

tensorflower-gardener committed:
Fix typo in export_savedmodel api doc.

PiperOrigin-RevId: 180203426

gunan committed:
Hack around a bug caused by bazel+tempfile+multiprocessing.

PiperOrigin-RevId: 180207077

gunan committed:
Bump the limit on checkpopint_utils_test as it is failing consistently with the current number by a small margin.

PiperOrigin-RevId: 180207280

rongjiecomputer committed:
[XLA] Replace GCC vector extension with portable Intel SIMD (#15553)

* [XLA] Replace GCC vector extension with portable Intel SIMD

* [XLA] Use #ifdef instead of #if for TF_XLA_HAS_*

yongtang committed:
Fix shape inference for bitwise ops with broadcasting (#14678)

* Fix shape inference for bitwise ops with broadcasting

This fix tries to address the issue raised in 14646 where
shape inference for bitwise ops is incorrect with broadcasting.
As was specified in 14646, in the following
```
>>> import tensorflow as tf
>>> tf.bitwise.bitwise_and(tf.zeros([3,1], dtype=tf.int32), tf.zeros([1,3], dtype=tf.int32))
<tf.Tensor 'BitwiseAnd:0' shape=(3, 1) dtype=int32>
```
the result shape should be (3, 3), not (3, 1).

This fix fixes the issue by changing
`.SetShapeFn(shape_inference::UnchangedShape)`
to
`.SetShapeFn(shape_inference::BroadcastBinaryOpShapeFn)`

This fix fixes 14646.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

* Add test cases for shape inference for bitwise ops with broadcasting

This commit adds test cases for shape inference of bitwise ops with broadcasting

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

* Update to add addtional assert for shape function

for comment feedback.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

im-minion committed:
updated Readme.md (#15666)

tf-nightly whl files added for Linux of python 3.6 version

isaprykin committed:
Improve a text comment in replicate_model_fn.GatheringOptimizer.

PiperOrigin-RevId: 180224227

zenofsahil committed:
Removed misplaced quote char (#15657)

macmonac committed:
Update learning.py (#14910)

use sys.maxsize because sys.maxint doesn't exist in Python 3

fdzzy committed:
Fix typo 'updaye' in audio_recognition.md (#15661)

themightyoarfish committed:
Fix small type in docstring which causes the API doc to render incorrectly (#15664)

Commit:
Merge commit for internal changes

sbl committed:
Predictor fixes for core estimators (#15648)

* fix predictor estimator factory

* fix checkpoint for chiefsessioncreator

Atlas7 committed:
minor wording fix in readme (#15667)

timefly666 committed:
support distributed training for esitmator api with multiple GPUs on the same machine (#14443)

Androbin committed:
Exclude tf_stream_executor.cmake for CPU only (#15600)

Bug introduced in #15099
Fixes #3996

caisq committed:
Merge pull request #15675 from av8ramit/branch_180224227

Branch 180224227

Commit:
Enabling tests to pass with python3.6. Updating dependencies for dock… (#15676)

* Enabling tests to pass with python3.6. Updating dependencies for docker tests.

* Removing an unnecessary update.

qmick committed:
Add name scope to tf.image (#15295)

* add name scope to tf.image

* improve name scope setting and format

* add tests

drpngx commented:
Sorry, you have to send this against master, not our release branch.
